homeassistant/components/wallbox/sensor.py

```diff
@@ -0,0 +1,93 @@
+"""Home Assistant component for accessing the Wallbox Portal API. The sensor component creates multiple sensors regarding wallbox performance."""
+
+from datetime import timedelta
+import logging
+
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.update_coordinator import (
+    CoordinatorEntity,
+    DataUpdateCoordinator,
+)
+
+from .const import CONF_CONNECTIONS, DOMAIN, SENSOR_TYPES
+
+CONF_STATION = "station"
+UPDATE_INTERVAL = 30
+
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def wallbox_updater(wallbox, hass):
+    """Get new sensor data for Wallbox component."""
+    data = await wallbox.async_get_data(hass)
+
+    filtered_data = {k: data[k] for k in SENSOR_TYPES if k in data}
+
+    for key, value in filtered_data.items():
+        sensor_round = SENSOR_TYPES[key]["ST_ROUND"]
+        if sensor_round:
+            try:
+                filtered_data[key] = round(value, sensor_round)
+            except TypeError:
+                _LOGGER.debug("Cannot format %s", key)
+
+    return filtered_data
+
+
+async def async_setup_entry(hass, config, async_add_entities):
+    """Create wallbox sensor entities in HASS."""
+    wallbox = hass.data[DOMAIN][CONF_CONNECTIONS][config.entry_id]
+
+    async def async_update_data():
+        return await wallbox_updater(wallbox, hass)
+
+    coordinator = DataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        # Name of the data. For logging purposes.
+        name="wallbox",
+        update_method=async_update_data,
+        # Polling interval. Will only be polled if there are subscribers.
+        update_interval=timedelta(seconds=UPDATE_INTERVAL),
+    )
+
+    await coordinator.async_refresh()
```

Comment:
```suggestion
    await coordinator.async_config_entry_first_refresh()
```

Reply:
But move this into `__init__.py` and store the coordinator in `hass.data`

Reply:
I had to google the 'first refresh' function, but that seems like a good addition. What would be the benefit of moving this to init? And where would you store the coordinator? Currently I store the wallbox object there, which has the benefit that I can include additional methods for future functionality

Reply:
Make hass.data[DOMAIN][entry.entry_id] as dict

Reply:
Thnx, I looked at the Roku implementation. I think that is quite a nice way to do it, but this would also mean a substantial rewrite (because I would also need a custom coordinator class if I want to use additional methods for the api). So I'm going to write this one down on my wishlist for a next big change if that's ok with you.

Reply:
Thats fine to not use the Roku method.  It still needs to be moved into `__init__` or Home Assistant won't be able to retry the setup since it needs to happen before the platforms are loaded

Reply:
Okay, I moved it around a bit. The coordinator is now being created in __init__ and becomes part of the WallboxHub class (which in turn is being stored in hass.data[DOMAIN][entry.entry_id]).